### PR TITLE
community[minor]: Revamp PGVector Filtering

### PR DESCRIPTION
**Pull request by eyurtsev**
This PR makes the following updates in the pgvector database:

1. Use JSONB field for metadata instead of JSON
2. Update operator syntax to include required `$` prefix before the operators (otherwise there will be name collisions with fields)
3. The change is non-breaking, old functionality is still the default, but it will emit a deprecation warning
4. Previous functionality has bugs associated with comparisons due to casting to text (so lexical ordering is used incorrectly for numeric fields)
5. Adds an a GIN index on the JSONB field for more efficient querying

### CODE REVIEW

File: libs/community/langchain_community/vectorstores/pgvector.py [hunk 1]
```diff
@@ -142,6 +210,20 @@ def _results_to_docs(docs_and_scores: Any) -> List[Document]:
     return [doc for doc, _ in docs_and_scores]
 
 
+def _sanitized_double_quoted_string(string: str) -> str:
```
Comment (eyurtsev) [inline libs/community/langchain_community/vectorstores/pgvector.py:213]: @nfcampos could you give this a second look -- this is dangerous code

**Comment by eyurtsev** [comment]
```

======================= 74 passed, 19 warnings in 3.02s ========================
PASSED                                   [  1%]PASSED                        [  2%]PASSED                    [  4%]PASSED        [  5%]PASSED                 [  6%]PASSED         [  8%]PASSED              [  9%]PASSED          [ 10%]PASSED                [ 12%]PASSED               [ 13%]PASSED                       [ 14%]PASSED                   [ 16%]PASSED        [ 17%]PASSED [ 18%]PASSED     [ 20%]PASSED [ 21%]PASSED            [ 22%]PASSED           [ 24%]PASSED [ 25%]PASSED [ 27%]PASSED [ 28%]PASSED [ 29%]PASSED [ 31%]PASSED [ 32%]PASSED [ 33%]PASSED [ 35%]PASSED [ 36%]PASSED [ 37%]PASSED [ 39%]PASSED [ 40%]PASSED [ 41%]PASSED [ 43%]PASSED [ 44%]PASSED [ 45%]PASSED [ 47%]PASSED [ 48%]PASSED [ 50%]PASSED [ 51%]PASSED [ 52%]PASSED [ 54%]PASSED [ 55%]PASSED [ 56%]PASSED [ 58%]PASSED [ 59%]PASSED [ 60%]PASSED [ 62%]PASSED [ 63%]PASSED [ 64%]PASSED [ 66%]PASSED [ 67%]PASSED [ 68%]PASSED [ 70%]PASSED [ 71%]PASSED           [ 72%]PASSED           [ 74%]PASSED           [ 75%]PASSED           [ 77%]PASSED           [ 78%]PASSED           [ 79%]PASSED              [ 81%]PASSED [ 82%]PASSED [ 83%]PASSED [ 85%]PASSED [ 86%]PASSED [ 87%]PASSED [ 89%]PASSED [ 90%]PASSED [ 91%]PASSED [ 93%]PASSED [ 94%]PASSED [ 95%]PASSED [ 97%]PASSED [ 98%]PASSED                         [100%]
Process finished with exit code 0
```

**Comment by eyurtsev** [comment]
Resolves issues by: 

https://github.com/langchain-ai/langchain/pull/12977
https://github.com/langchain-ai/langchain/pull/11897
https://github.com/langchain-ai/langchain/pull/17067

File: libs/community/langchain_community/vectorstores/pgvector.py [hunk 2]
```diff
@@ -209,6 +306,29 @@ def __init__(
         self.override_relevance_score_fn = relevance_score_fn
         self.engine_args = engine_args or {}
         self._bind = connection if connection else self._create_engine()
+        self.use_jsonb = use_jsonb
+
+        if not use_jsonb:
+            # Replace with a deprecation warning.
+            warn_deprecated(
+                "0.0.27",
```
Comment (baskaryan) [inline libs/community/langchain_community/vectorstores/pgvector.py:314]: 0.0.29

File: libs/community/langchain_community/vectorstores/pgvector.py [hunk 3]
```diff
@@ -508,7 +631,134 @@ def _results_to_docs_and_scores(self, results: Any) -> List[Tuple[Document, floa
...
+                path_value = sqlalchemy.text(
+                    f"'$.{field} {native} {sanitized_value}'",
+                )
+            else:
+                placeholder_name = f"filter_value_{counter.increment()}"
+                path_value = sqlalchemy.text(
+                    f"'$.{field} {native} :{placeholder_name}'",
+                ).bindparams(**{placeholder_name: filter_value})
+            return json_path_operation(path_value)
+        elif operator in "$between":
```
Comment (baskaryan) [inline libs/community/langchain_community/vectorstores/pgvector.py:713]: equality?

File: libs/community/langchain_community/vectorstores/pgvector.py [hunk 4]
```diff
@@ -508,7 +631,134 @@ def _results_to_docs_and_scores(self, results: Any) -> List[Tuple[Document, floa
...
+            # Use AND with two comparisons
+            low, high = filter_value
+            placeholder_name = f"value_{counter.increment()}"
+            lower_bound = sqlalchemy.text(
+                f"'$.{field} >= :{placeholder_name}'",
+            ).bindparams(**{placeholder_name: low})
+
+            placeholder_name = f"value_{counter.increment()}"
+            upper_bound = sqlalchemy.text(
+                f"'$.{field} <= :{placeholder_name}'",
```
Comment (baskaryan) [inline libs/community/langchain_community/vectorstores/pgvector.py:723]: is it standard for between to mean inclusive on both sides

Comment (eyurtsev) [reply libs/community/langchain_community/vectorstores/pgvector.py:723]: I think that's the convention in SQL

File: libs/community/langchain_community/vectorstores/pgvector.py (hunk 3 above)
Comment (eyurtsev) [reply libs/community/langchain_community/vectorstores/pgvector.py:677]: equality is handled by line above which handles all comparison operations

Comment (baskaryan) [reply libs/community/langchain_community/vectorstores/pgvector.py:673]: as in shoudl this be `elif operator == "$between"`

**Review by baskaryan — APPROVED** [review]

**Comment by mlucool** [comment]
Hi.

As noted in https://github.com/langchain-ai/langchain/issues/19681, this is a breaking change. Can this be made non breaking from a sqlalchemy POV? Today langchian [declares ](https://github.com/langchain-ai/langchain/blob/master/libs/langchain/pyproject.toml#L20)`SQLAlchemy = ">=1.4,<3"`

Can this be made backwards compatible and/or reverted?

**Pull request merged**